cpp/tribune-client/main.cpp

```diff
@@ -1,74 +1,246 @@
-        std::cerr << e.what() << std::endl;
-        std::cout << optionsDescription;
-        return 1;
-    }
+		std::string voiceName = userOptions["voice-name"].as<std::string>();
+		std::string voiceGender = userOptions["voice-gender"].as<std::string>();
+		std::string voiceAge = userOptions["voice-age"].as<std::string>();
+		if (!voiceName.empty() || !voiceGender.empty() || !voiceAge.empty())
+		{
+			synthesizeConfig.voice.emplace();
+			synthesizeConfig.voice->name = voiceName;
+			synthesizeConfig.voice->gender = genderFromString(voiceGender);
+			synthesizeConfig.voice->age = ageFromString(voiceAge);
+		}
 
-    try {
-        const auto sample_rate_hertz = userOptions["sample-rate-hertz"].as<unsigned int>();
+		std::string outputPath = userOptions["out-path"].as<std::string>();
+		if (outputPath.empty())
+		{
+			if (encoding == techmo::tribune::AudioEncoding::PCM16)
+			{
+				outputPath = "TechmoTTS.wav";
+			}
+			else if (encoding == techmo::tribune::AudioEncoding::OGG_VORBIS)
+			{
+				outputPath = "TechmoTTS.ogg";
+			}
+		}
 
-        techmo::tribune::TribuneClientConfig config;
-        config.session_id = userOptions["session-id"].as<std::string>();
-        config.grpc_timeout = userOptions["grpc-timeout"].as<int>();
-        config.sample_rate_hertz = sample_rate_hertz;
+		bool useStreaming;
+		if (userOptions["response"].as<std::string>() == "streaming")
+		{
+			useStreaming = true;
+		}
+		else if (userOptions["response"].as<std::string>() == "single")
+		{
+			useStreaming = false;
+		}
+		else
+		{
+			std::string message{ "Unsupported response: " };
+			message += userOptions["response"].as<std::string>();
+			message += ".";
+			throw std::runtime_error{ message };
+		}
 
-        techmo::tribune::TribuneClient tribune_client{ userOptions["service-address"].as<std::string>() };
+		std::string requestText = userOptions["text"].as<std::string>();
+		const auto audio_data = (useStreaming ?
+			tribuneClient.SynthesizeStreaming(clientConfig, synthesizeConfig, requestText) :
+			tribuneClient.Synthesize(clientConfig, synthesizeConfig, requestText));
 
-        const auto audio_data = tribune_client.Synthesize(config, userOptions["text"].as<std::string>());
+		if (encoding == techmo::tribune::AudioEncoding::PCM16)
+		{
+			WriteWaveFile(outputPath, audio_data.sample_rate_hertz, audio_data.audio_bytes);
+		}
+		else
+		{
+			WriteEncodedFile(outputPath, audio_data.audio_bytes);
+		}
+	}
+	catch (const std::exception& e)
+	{
+		std::cerr << e.what() << std::endl;
+		return 1;
+	}
+
+	return 0;
+}
+
+void printVoices(const std::vector<techmo::tribune::SynthesizeVoiceInfo>& voices)
+{
+	std::cout << std::endl << "Available voices:" << std::endl << std::endl;
+	for (const auto& voice: voices)
+	{
+		std::cout << "name: '" << voice.voice.name << "'" << std::endl;
+		std::cout << "gender: " << voice.voice.gender << std::endl;
+		std::cout << "age: " << voice.voice.age << std::endl << std::endl;
```

Comment:
supported languages could also be printed here

Reply:
and it would be better to print names of gender/age instead of ints

Reply:
Functionality added.